### DISCUSSION TITLE
Change Number formatter throught the application

### DISCUSSION THREAD

**Discussion by EdwinThomas11** [serenity-is/Serenity#7038] [Help] [answered]
export class CustomNumberFormatter extends Serenity.NumberFormatter {

        private tenantSettings : TenantSettingsRow;
        constructor() {
            super();
	       this.tenantSettings = TenantSettingsRow.getLookup().items.filter((x) => x.TenantId === tenantId)[0];
        }

        public format(ctx: Slick.FormatterContext): string {           
            // this.displayFormat =  "#.##0,00";
            this.displayFormat =  "#.##0,00";
            return super.format(ctx);
        }
    }
    
   
I must modify the number format throughout my application based on specific settings. For instance, formats like 1.234,56 and 234,567.89 are required, aligning with international standards where dots are used in place of decimals.

However, I'm encountering difficulty achieving the desired format "1.234,56" when using "#.##0,00". Assistance in resolving this matter would be greatly appreciated. @VictorTomaili @dfaruque @volkanceylan

**Comment by EdwinThomas11** [comment]
I would need the same format in columns and forms,

**Comment by EdwinThomas11** [comment]
I've extended `Slick.Formatter`, and while it functions as intended in columns, it is currently not functioning as expected in forms.

please find the code used for reference:

namespace DPWBooking.Default {

    @Serenity.Decorators.registerFormatter()
    export class CustomNumberFormatter implements Slick.Formatter {

        public format(ctx: Slick.FormatterContext): string {
            if (ctx.value == null) {
                return '';
            }
            var text = Q.formatNumber(ctx.value, "#.##0,00"); // different number formats based on settings.
            return text;
        }
    }
}

**Comment by EdwinThomas11** [comment]
@VictorTomaili @dfaruque appreciate your inputs here.

**Comment by VictorTomaili** [comment, marked as answer]
You can use `Culture.decimalSeparator` in your format. This value is changed based user culture.

**Reply by EdwinThomas11** [reply]
This functions flawlessly, precisely as I desired. Many thanks, @VictorTomaili . Your assistance is highly valued.

**Comment by EdwinThomas11** [comment]
@VictorTomaili Can we set a different decimal separator specifically for one or two decimal fields? If possible, could you please provide guidance on how to achieve this?

**Reply by VictorTomaili** [reply]
You can use anything instead of `Culture.decimalSeparator`

**Reply by EdwinThomas11** [reply]
@VictorTomaili Thanks for the quick response. I require assistance in configuring the display format for group separators across my entire application, based on my specified settings. For example, I want to determine whether the format should be like "11,00,000" or "1,100,000," with commas representing either hundreds or thousands. Is there a global setting or option to define this format?

**Reply by VictorTomaili** [reply]
There is no global settings or option for this. You may override the Culture options, or you may use custom formatters editors etc. We didn't have a sample for overriding culture options, but I think you can do it in ScriptInit.ts file.

**Reply by EdwinThomas11** [reply]
var CustomNumberFormatter = Serenity.NumberFormatter;
  Serenity.NumberFormatter.prototype.format = function (ctx) {
    var format = "Your Format Here";
    }
    return CustomNumberFormatter.format(ctx.value, format);
  };
  
The mentioned code in scriptInit.ts functioned well in grids but is not working for forms. Using a custom attribute for each property solves the problem, but is there a way to have a solution similar to grids? @VictorTomaili

**Reply by VictorTomaili** [reply]
You can override `get_value` `set_value` functions for decimal editors.

**Reply by EdwinThomas11** [reply]
@VictorTomaili I attempted to override this previously, and although the `set_value` function is being triggered upon loading, it fails to trigger when the user modifies the values. As a result, I am unable to format the decimal field upon its change.
Also, the set_value always accept the value as numbers.

**Reply by VictorTomaili** [reply]
write your own editor to get full control for it. You can write a property processor to set default editor type as your custom editor for decimal or float properties.

**Reply by EdwinThomas11** [reply]
@VictorTomaili I've customised the defaultAutoNumericOptions of DecimalEditor, specifically adjusting the dGroup parameter from 3 to 2 through an override. This adjustment, along with a few tweaks in Q.validateOptions within the normalizer, is working well.

I trust that this will fulfil the requirement. Thank you very much for your assistance. @VictorTomaili.
